### DISCUSSION TITLE
Support Entity Relationships?

### DISCUSSION THREAD

**Discussion by graphnode** [genaray/Arch#96] [Ideas]
Support Entity Relationships like demonstrated here: https://ajmmertens.medium.com/building-games-in-ecs-with-entity-relationships-657275ba2c6c

This would allow to easily do hierarchies and other structured data.

**Comment by genaray** [comment]
So this feature was requested multiple times. I think theres actually enough demand for that ^^ 

I only see one major problem... This is c# and not c/c++ or rust. So we run into one huge problem, no advanced compile time statics. 

In flecs, if i saw that correctly, the entity relation queries are using some sort of additional descriptive meta language ( "$this") to describe what exactly is required. Thats not possible in csharp yet. 

While we could easily add relations, the query part would be a nightmare. So either source generation comes in handy here or some manual work. 

For now this can be "simulated" by using generics or just plain arrays in components :> 

```csharp
entity.Add<(ChildOf,Parent)>();
entity.Get<ParentOf>().Childs.Add(...);
```

**Reply by Mathetis** [reply]
`entity.Add<(ChildOf,Parent)>();`

Are you sure this works? I think the tuple results in an exception:

System.ArgumentException : The specified Type must not be a generic type. (Parameter 't')

**Comment by genaray** [comment]
Done with : https://github.com/genaray/Arch.Extended/wiki/Relationships
